File: goblin-api/src/main/java/org/goblinframework/api/rpc/ServiceTimeout.java
package org.goblinframework.api.rpc;

import java.lang.annotation.*;
import java.util.concurrent.TimeUnit;

@Documented
@Target({ElementType.TYPE, ElementType.METHOD, ElementType.FIELD})
@Retention(RetentionPolicy.RUNTIME)
public @interface ServiceTimeout {

  int timeout();

  TimeUnit unit();

  boolean enable() default true;

}

File: goblin-rpc/src/main/java/org/goblinframework/rpc/service/_ServiceTimeout.kt
package org.goblinframework.rpc.service

import org.goblinframework.api.rpc.ServiceTimeout

fun calculateServiceTimeout(annotation: ServiceTimeout?): Long? {
  return annotation?.run {
    if (this.enable) {
      val t: Int = this.timeout
      if (t < 0) {
        null
      } else {
        val timeout: Long = this.unit.toMillis(t.toLong())
        return if (timeout < 0) null else timeout
      }
    } else {
      null
    }
  }
}